fido2/hid/windows.py

```diff
@@ -0,0 +1,345 @@
+    wintypes.DWORD,
+    wintypes.DWORD,
+    ctypes.c_void_p,
+    wintypes.DWORD,
+    wintypes.DWORD,
+    HANDLE,
+]
+kernel32.CloseHandle.restype = wintypes.BOOL
+kernel32.CloseHandle.argtypes = [HANDLE]
+
+
+class WinCtapHidConnection(CtapHidConnection):
+    def __init__(self, descriptor):
+        self.descriptor = descriptor
+        self.handle = kernel32.CreateFileA(
+            descriptor.path,
+            GENERIC_WRITE | GENERIC_READ,
+            FILE_SHARE_READ | FILE_SHARE_WRITE,
+            None,
+            OPEN_EXISTING,
+            0,
+            None,
+        )
+        if self.handle == INVALID_HANDLE_VALUE:
+            raise ctypes.WinError()
+
+    def close(self):
+        kernel32.CloseHandle(self.handle)
+
+    def write_packet(self, packet):
+        out = b"\0" + packet  # Prepend report ID
+        num_written = wintypes.DWORD()
+        ret = kernel32.WriteFile(
+            self.handle, out, len(out), ctypes.byref(num_written), None
+        )
+        if not ret:
+            raise ctypes.WinError()
+        if num_written.value != len(out):
+            raise OSError(
+                "Failed to write complete packet.  "
+                + "Expected %d, but got %d" % (len(out), num_written.value)
+            )
+
+    def read_packet(self):
+        buf = ctypes.create_string_buffer(self.descriptor.report_size_in + 1)
+        num_read = wintypes.DWORD()
+        ret = kernel32.ReadFile(
+            self.handle, buf, len(buf), ctypes.byref(num_read), None
+        )
+        if not ret:
+            raise ctypes.WinError()
+
+        if num_read.value != self.descriptor.report_size_in + 1:
+            raise OSError("Failed to read full length report from device.")
+
+        return buf.raw[1:]  # Strip report ID
+
+
+def get_vid_pid(device):
+    attributes = HidAttributes()
+    result = hid.HidD_GetAttributes(device, ctypes.byref(attributes))
+    if not result:
+        raise ctypes.WinError()
+
+    return attributes.VendorID, attributes.ProductID
+
+
+def get_descriptor(path):
+    device = kernel32.CreateFileA(
+        path, 0, FILE_SHARE_READ | FILE_SHARE_WRITE, None, OPEN_EXISTING, 0, None,
+    )
+    if device == INVALID_HANDLE_VALUE:
+        raise ctypes.WinError()
+    try:
+        preparsed_data = PHIDP_PREPARSED_DATA(0)
+        ret = hid.HidD_GetPreparsedData(device, ctypes.byref(preparsed_data))
+        if not ret:
+            raise ctypes.WinError()
+
+        try:
+            caps = HidCapabilities()
+            ret = hid.HidP_GetCaps(preparsed_data, ctypes.byref(caps))
+
+            if ret != HIDP_STATUS_SUCCESS:
+                raise ctypes.WinError()
+
+            if caps.UsagePage == FIDO_USAGE_PAGE and caps.Usage == FIDO_USAGE:
+                vid, pid = get_vid_pid(device)
+                size_in = caps.InputReportByteLength - 1
+                size_out = caps.OutputReportByteLength - 1
+                return HidDescriptor(path, vid, pid, size_in, size_out)
```

Comment:
These - 1 are in the old code as well but I do not get them. These are not null terminated strings. Have we looked closer into this? Maybe not relevant for this PR though.

Reply:
I'll add a comment to clarify this. As per https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/hidpi/ns-hidpi-_hidp_caps these sizes include the 1-byte report ID which is pre-pended to the payload.